### ISSUE TITLE
Implement basic tests for views

### ISSUE THREAD

**Issue by philipbelesky** [czlee/tabbycat#319]
As I'm going through some early steps of #297 I'm also making decent progress on #228. At the point most of our views are class based, it would make sense to add some basic test cases which would basically just test 'does this view crash'.

I'm not super familiar with the testing framework, but to do so I'm imaging we would load a fixture from a fake completed tournament and then query each view and check it returns something without throwing an exception. Not the most advanced test case, but it would hopefully catch a fair few regressions.

**Comment by czlee** [comment]
Seems sensible, I can look into this some time if you want? I'm sure there are some good resources for this, we can't be the first Django users to need to do this.

Ideally we'd want this to require minimal (or zero) maintenance I guess, so that it would auto-discover views and just test all of them. Though, a few views are designed to throw `Http404()` exceptions, so we'd need to allow for that.

**Comment by czlee** [comment]
https://docs.djangoproject.com/en/1.9/intro/tutorial05/#test-a-view

**Comment by czlee** [comment]
Observations:
 1. The views testing client doesn't care whether or not views are class-based; Django has an in-built test client.
 2. Writing more specific test cases doesn't seem that hard, maybe I take back what I said about auto-discovery being ideal. On reflection I'm also just not sure whether auto-discovery is sensible, since there are specifics relating to logins and redirects and the like that could get in the way.
 3. I suspect we'll need `TestCase` subclasses that can deal efficiently with round and tournament views, or decorators that will run a test method on each round or something.

**Comment by philipbelesky** [comment]
Sounds good. I guess also there is a question in terms of how to divide up test responsibility. For example in a team tab view, should the view's test check that the each of the rankings (as rendered into the table) are correct? Or is that something that should only be exclusively testing in the dedicated tests for standings?

**Comment by czlee** [comment]
Yeah, that's a great question. My instinct is the latter, since if you change how standings work (for whatever reason), you don't want to have to change the test for every view that uses it.

**Comment by philipbelesky** [comment]
Yea. To flesh the example out: the majority of our views can be expressed as "dump this data to a table and give it a header". In converting the tables over to a standardised vue component we can in many cases get rid of page specific templates entirely by passing the table data (as a json dump) and page title to a generic template.

In these cases there are a number of relatively easy tests than can be written against the json dump that do not intrude (too much) on the business logic of what is being presented — such as:

- Does this table have any data (if that is expected)
- Do the number of rows in the data set match what is expected (number of speakers at the tournament, number of active teams / 2)
- Are particular keys within that data set present (ie "Institution" if `show_institutions` is enabled).

Any other ideas? Most view bugs seem to come down to either:

- This design element is broken or styled poorly (in which case tests can't do a lot to help here).
- Something in the view logic or underlying business logic is causing an exception (in which case most of the tests proposed above are somewhat unnecessary)

Which would lead me to this strategy

- Provide a fixture and auto discover all views (with perhaps additional logic for base classes such as `TournamentMixin` and `RoundMixin` to check nothing is throwing an exception (fixture should have a data set such that no view throws an exceptions with intended exceptions caveated from this test. 
- For views which may or may not have data or may or throw deliberate exceptions depending on tournament state (ie break pages or draw pages) write some special tests that simulate and check these cases. Do a similar thing for views that have variables keys or data of a variable length. 
- As we come across particular bugs that arise given particular states maybe add those in as we come across them? I'm struggling to think of any examples here that wouldn't be captured by the above cases.

**Comment by czlee** [comment]
I don't really have a good enough understanding of Vue to fully understand this, but yeah, this makes sense. I guess for very basic tests this would work and be useful. We'd need to set a few base assumptions, like:
* Requests will always be "in scope", _e.g._ not for a silent round
* User will always be logged in

I think realistically autodiscover won't work very well, since you'd have to differentiate between pages that are tables (or at least pages that are not redirects), and pages that aren't automatically testable. Though... now that I think of it, if you have a base class for the new table view, then I guess you'd just look for all subclasses of it, and that will differentiate between pages just fine?

Would the generic table work for things like team standings, the template logic for which is currently quite complicated? I guess you'd just move all the logic to Python? I spent a while designing the `Standings` class to be Django-template-accessible (lol).

**Comment by philipbelesky** [comment]
Hmm, for the "doesnt throw an exception" test that could plausibly autodiscover almost all views, but yes for anything table data related it would limit itself to views inheriting from the base class.

I'm currently converting some of the standings pages over, and yes the logic is being moved over to python. Views end up a bit larger, but we should be able to remove approximately half of the current templates, and make tables much easier to build and consistently presented. I might ping you at some point about how to improve the abstractions here once I flesh out the last details around accomodating  different types of cell content (links; hovers) and standardising header titles (having consistent icons/hover titles etc).
